### PR TITLE
stylo: Error reporting for unknown media features

### PR DESCRIPTION
**Pull request by ferjm**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [bug 1384225](https://bugzilla.mozilla.org/show_bug.cgi?id=1384225)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18341)

### CODE REVIEW

File: components/style/media_queries.rs [hunk 1]
```diff
@@ -240,19 +241,29 @@ impl MediaQuery {
 /// media query list is only filled with the equivalent of "not all", see:
 ///
 /// https://drafts.csswg.org/mediaqueries/#error-handling
-pub fn parse_media_query_list(context: &ParserContext, input: &mut Parser) -> MediaList {
+pub fn parse_media_query_list<R>(context: &ParserContext,
```
Comment (emilio) [inline components/style/media_queries.rs]: Let's format this as rustfmt would:

```rust
pub fn parse_media_query_list<R>(
    context: &ParserContext,
    input: &mut Parser,
    error_reporter: &R,
) -> MediaList
where
    R: ParseErrorReporter,
{
```

File: components/style/media_queries.rs [hunk 2]
```diff
@@ -240,19 +241,29 @@ impl MediaQuery {
...
+        let start_position = input.position();
+        let start_location = input.current_source_location();
         match input.parse_until_before(Delimiter::Comma, |i| MediaQuery::parse(context, i)) {
             Ok(mq) => {
                 media_queries.push(mq);
             },
-            Err(..) => {
+            Err(err) => {
                 media_queries.push(MediaQuery::never_matching());
+                let error = ContextualParseError::InvalidMediaFeature(
```
Comment (emilio) [inline components/style/media_queries.rs]: So this would report `Expected media feature name but found...`, but there are definitely more errors than can happen instead... What does gecko do for those? Does it report them? I suspect not.

Comment (jdm) [reply components/style/media_queries.rs]: Yes; we will want to add a new variant to StyleParseError that is only returned when an invalid media feature is encountered, and only report if that variant is encountered here.

**Comment by jdm** [comment]
Can you add a new test to https://dxr.mozilla.org/mozilla-central/source/layout/style/test/test_css_parse_error_smoketest.html for this as well?

**Comment by ferjm** [comment]
Feedback addressed and test added https://bugzilla.mozilla.org/show_bug.cgi?id=1384225#c5

File: components/style/media_queries.rs [hunk 3]
```diff
@@ -240,19 +242,43 @@ impl MediaQuery {
...
+                            StyleParseError::MediaQueryExpectedFeatureName)) =>
+                        ContextualParseError::MediaQueryExpectedFeatureName(
+                            input.slice_from(start_position), err),
+                    CssParseError::Custom(SelectorParseError::Custom(
+                            StyleParseError::RangedExpressionWithNoValue)) =>
+                        ContextualParseError::MediaQueryNoMinMaxWithoutValue,
+                    CssParseError::Custom(SelectorParseError::Custom(
+                            StyleParseError::MediaQueryExpectedFeatureValue)) =>
+                        ContextualParseError::MediaQueryExpectedFeatureValue,
+                    _ => ContextualParseError::UnknownAtRule(input.slice_from(start_position), err),
```
Comment (emilio) [inline components/style/media_queries.rs]: `UnknownAtRule` is not necessarily correct, is it? We can try to parse media query lists from a whole set of different places, like `@import` rules, stylesheet attributes...

File: components/style/gecko/media_queries.rs [hunk 4]
```diff
@@ -530,8 +530,9 @@ impl Expression {
                     Ok((f, r)) => {
                         feature = f;
                         range = r;
-                    }
-                    Err(()) => return Err(SelectorParseError::UnexpectedIdent(ident.clone()).into()),
+                    },
+                    Err(()) =>
```
Comment (emilio) [inline components/style/gecko/media_queries.rs]: nit: use a block here.

File: components/style/media_queries.rs (hunk 3 above)
Comment (ferjm) [reply components/style/media_queries.rs]: Ah, right. I guess something like a new `UnknownMediaQuery` will do then?

Comment (jdm) [reply components/style/media_queries.rs]: Rather than matching on known errors here, we should simply have a `InvalidMediaRule(ParseError)` variant.

File: ports/geckolib/error_reporter.rs [hunk 5]
```diff
@@ -281,6 +285,12 @@ impl<'a> ErrorHelpers<'a> for ContextualParseError<'a> {
             }
             ContextualParseError::UnsupportedRule(..) =>
                 (b"PEDeclDropped\0", Action::Nothing),
+            ContextualParseError::MediaQueryExpectedFeatureName(..) =>
+                (b"PEMQExpectedFeatureName\0", Action::Nothing),
+            ContextualParseError::MediaQueryExpectedFeatureValue =>
+                (b"PEMQExpectedFeatureValue\0", Action::Nothing),
+            ContextualParseError::MediaQueryNoMinMaxWithoutValue =>
+                (b"PEMQNoMinMaxWithoutValue\0", Action::Nothing),
```
Comment (jdm) [inline ports/geckolib/error_reporter.rs]: These will look like:
```rust
ContextualPArseError::InvalidMediaQuery(
    CssParseError::Custom(SelectorParseError::Custom(StyleParseError::MediaQueryExpectedFeatureName(..)))) =>
    (b"PEMQExpectedFeatureName\0", Action::Nothing),
```

**Review by jdm — CHANGES_REQUESTED** [review]
Other testcase:
`<style>@media "foo" {}</style>` should yield `Expected identifier in media list but found ‘"foo"’.`
The use of RangedExpressionWithNoValue should also be an instance of MediaQueryExpectedFeatureValue, I believe. The use of UnexpectedIdent should be replaced by MediaQueryExpectedFeatureName instead.

File: components/style/gecko/media_queries.rs [hunk 6]
```diff
@@ -530,17 +616,19 @@ impl Expression {
                     Ok((f, r)) => {
                         feature = f;
                         range = r;
-                    }
-                    Err(()) => return Err(SelectorParseError::UnexpectedIdent(ident.clone()).into()),
+                    },
+                    Err(()) => {
+                        return Err(StyleParseError::MediaQueryExpectedFeatureName.into())
```
Comment (jdm) [inline components/style/gecko/media_queries.rs]: We should pass the ident that was not recognized as a parameter of this error variant.

File: components/style/gecko/media_queries.rs [hunk 7]
```diff
@@ -488,13 +570,17 @@ impl Expression {
...
+            StyleParseError::MediaQueryExpectedIdentifier
+        )?;
         input.parse_nested_block(|input| {
             // FIXME: remove extra indented block when lifetimes are non-lexical
             let feature;
             let range;
             {
-                let ident = input.expect_ident()?;
+                let ident = input.expect_ident().map_err(|_|
+                    StyleParseError::MediaQueryExpectedFeatureName
```
Comment (jdm) [inline components/style/gecko/media_queries.rs]: This should be ExpectedIdentifier, instead. We can match an UnexpectedToken error, extract it and add it as a parameter of the error variant.

File: components/style/gecko/media_queries.rs [hunk 8]
```diff
@@ -488,13 +570,17 @@ impl Expression {
     /// ```
     pub fn parse<'i, 't>(context: &ParserContext, input: &mut Parser<'i, 't>)
                          -> Result<Self, ParseError<'i>> {
-        input.expect_parenthesis_block()?;
+        input.expect_parenthesis_block().map_err(|_|
```
Comment (jdm) [inline components/style/gecko/media_queries.rs]: Gecko does not report an error here, so we can revert this change.

**Review by jdm — CHANGES_REQUESTED** [review]
Almost there! We can get better error messages by explicitly adding the token or identifier to the error variants, however.

File: components/style/gecko/media_queries.rs (hunk 8 above)
Comment (ferjm) [reply components/style/gecko/media_queries.rs]: If I revert this, `@media "foo" {}` would only report "Declaration dropped" instead of "Expected identifier in media list but found ‘ "foo" ’"

**Comment by jdm** [comment]
@bors-servo: r+
Thanks!

**Comment by ferjm** [comment]
@bors-servo: retry

I think these are intermittents

**Comment by KiChjang** [comment]
@ferjm I'm not in front of a computer, could you please open new issues filing the intermittents you see here?

**Comment by jdm** [comment]
#18424 and #18273
